contrib/golang/filters/http/source/golang_filter.h

```diff
@@ -0,0 +1,258 @@
+
+  static std::atomic<uint64_t> global_stream_id_;
+
+  void continueStatus(GolangStatus status);
+
+  void sendLocalReply(Http::Code response_code, absl::string_view body_text,
+                      std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+                      Grpc::Status::GrpcStatus grpc_status, absl::string_view details);
+
+  absl::optional<absl::string_view> getHeader(absl::string_view key);
+  void copyHeaders(GoString* go_strs, char* go_buf);
+  void setHeader(absl::string_view key, absl::string_view value);
+  void removeHeader(absl::string_view key);
+  void copyBuffer(Buffer::Instance* buffer, char* data);
+  void setBufferHelper(Buffer::Instance* buffer, absl::string_view& value, bufferAction action);
+  void copyTrailers(GoString* go_strs, char* go_buf);
+  void setTrailer(absl::string_view key, absl::string_view value);
+  void getStringValue(int id, GoString* value_str);
+
+private:
+  ProcessorState& getProcessorState();
+
+  bool doHeaders(ProcessorState& state, Http::RequestOrResponseHeaderMap& headers, bool end_stream);
+  GolangStatus doHeadersGo(ProcessorState& state, Http::RequestOrResponseHeaderMap& headers,
+                           bool end_stream);
+  bool doData(ProcessorState& state, Buffer::Instance&, bool);
+  bool doDataGo(ProcessorState& state, Buffer::Instance& data, bool end_stream);
+  bool doTrailer(ProcessorState& state, Http::HeaderMap& trailers);
+  bool doTrailerGo(ProcessorState& state, Http::HeaderMap& trailers);
+
+  uint64_t getMergedConfigId(ProcessorState& state);
+
+  void continueEncodeLocalReply(ProcessorState& state);
+  void continueStatusInternal(GolangStatus status);
+  void continueData(ProcessorState& state);
+
+  void onHeadersModified();
+
+  void sendLocalReplyInternal(Http::Code response_code, absl::string_view body_text,
+                              std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+                              Grpc::Status::GrpcStatus grpc_status, absl::string_view details);
+
+  const FilterConfigSharedPtr config_;
+  Dso::DsoInstancePtr dynamic_lib_;
+
+  Http::RequestOrResponseHeaderMap* headers_{nullptr};
+  Http::HeaderMap* trailers_{nullptr};
+
+  // save temp values from local reply
+  Http::RequestOrResponseHeaderMap* local_headers_{nullptr};
+  Http::HeaderMap* local_trailers_{nullptr};
+
+  // The state of the filter on both the encoding and decoding side.
+  DecodingProcessorState decoding_state_;
+  EncodingProcessorState encoding_state_;
+
+  // TODO get all context
+  Grpc::Context& context_;
+  uint64_t cost_time_decode_{0};
+  uint64_t cost_time_encode_{0};
+  uint64_t cost_time_mem_{0};
+  uint64_t stream_id_{0};
+
+  httpRequestInternal* req_{nullptr};
+
+  // lock for has_destroyed_,
+  // to avoid race between envoy c thread and go thread (when calling back from go).
+  // it should also be okay without this lock in most cases, just for extreme case.
+  Thread::MutexBasicLockable mutex_{};
```

Comment:
Document which variables are covered by the lock via ABSL_EXCLUSIVE_LOCKS_REQUIRED and similar. I believe various headers also have to be covered by this due to how you are not posting all modifications for headers gets/sets/etc.

Reply:
Done.